runtime/src/bank.rs

```diff
@@ -626,12 +627,73 @@ impl Bank {
             Ordering::Relaxed,
         );
     }
 
+    /// iterate over all stakes, redeem vote credits for each stake we can
+    ///   successfully load and parse, return total payout
+    fn pay_validator_rewards(&self, validator_point_value: f64) -> u64 {
+        let stake_history = self.stakes.read().unwrap().history().clone();
+        self.stake_delegations()
+            .iter()
+            .map(
+                |(
+                    stake_pubkey,
+                    Delegation {
+                        voter_pubkey: vote_pubkey,
+                        ..
+                    },
+                )| {
+                    match (
+                        self.get_account(&stake_pubkey),
+                        self.get_account(&vote_pubkey),
+                    ) {
+                        (Some(mut stake_account), Some(mut vote_account)) => {
+                            match (stake_account.state(), vote_account.state()) {
+                                (Ok(StakeState::Stake(meta, mut stake)), Ok(vote_state)) => {
+                                    // we've recovered everything we need to do redemption at this point
+                                    if let Some((voters_reward, stakers_reward, credits_observed)) =
+                                        stake.calculate_rewards(
+                                            validator_point_value,
+                                            &vote_state,
+                                            Some(&stake_history),
+                                        )
+                                    {
+                                        stake_account.lamports += stakers_reward;
+                                        vote_account.lamports += voters_reward;
+
+                                        stake.credits_observed = credits_observed;
+                                        stake.delegation.stake += stakers_reward;
+
+                                        if stake_account
+                                            .set_state(&StakeState::Stake(meta, stake))
+                                            .is_ok()
+                                        {
+                                            self.store_account(&stake_pubkey, &stake_account);
+                                            self.store_account(&vote_pubkey, &vote_account);
+                                            stakers_reward + voters_reward
+                                        } else {
+                                            0
+                                        }
+                                    } else {
+                                        0
+                                    }
+                                }
+                                (_, _) => 0,
+                            }
+                        }
+                        (_, _) => 0,
+                    }
+                },
+            )
+            .sum()
+    }
```

Comment:
tests?

Reply:
much of testing is covered by `test_bank_update_rewards()`.  bit that's missing is balance update verifications

Reply:
Is it though? This thing has a whole lot of branches.

Reply:
it has essentially 1 branch: whether or not there's meaningful rewards to redeem (tested to death in the stake program).  All the corner cases are there to handle the types and not abort.

maybe I should replace all the matches with simple `unwrap()`s/`expect()`s.  there are no valid cases where information coming out of bank.stakes can disagree with what's in the accounts_db.